### PR TITLE
fix(ivy): `ViewRef.rootNodes` returns rootNodes from child views as well

### PR DESCRIPTION
**Pull request by mhevery**
The idea behind `rootNodes` is to return all of the nodes which need
to be added to the DOM in order for the `ViewRef` to be rendered.
The new implementation correctly takes into account the case where
one of the rootNodes is a container containing other `ViewRef`s.

### CODE REVIEW

File: packages/core/src/render3/view_ref.ts [hunk 1]
```diff
@@ -299,27 +301,35 @@ export class RootViewRef<T> extends ViewRef<T> {
...
-      collectNativeNodes(lView, tNodeChild, result);
-    } else if (tNodeChild.type === TNodeType.Projection) {
+    if (tNode.type === TNodeType.ElementContainer) {
+      collectNativeNodes(lView, tNode.child, result);
+    } else if (tNode.type === TNodeType.Container) {
+      const lContainer = lView[tNode.index];
+      const containerTView = tNode.tViews !as TView;
+      ngDevMode && assertDefined(containerTView, 'TView expected');
+      const containerFirstChildTNode = containerTView.firstChild !;
+      if (containerFirstChildTNode) {
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/view_ref.ts]: Please use a stricter check here (`===` for `null` or whatever the falsy value might be).

File: packages/core/src/render3/view_ref.ts [hunk 2]
```diff
@@ -299,27 +301,35 @@ export class RootViewRef<T> extends ViewRef<T> {
...
-  while (tNodeChild) {
-    const nativeNode = getNativeByTNodeOrNull(tNodeChild, lView);
+function collectNativeNodes(lView: LView, tNode: TNode | null, result: any[]): any[] {
+  while (tNode) {
+    const nativeNode = getNativeByTNodeOrNull(tNode, lView);
     nativeNode && result.push(nativeNode);
-    if (tNodeChild.type === TNodeType.ElementContainer) {
-      collectNativeNodes(lView, tNodeChild, result);
-    } else if (tNodeChild.type === TNodeType.Projection) {
+    if (tNode.type === TNodeType.ElementContainer) {
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/view_ref.ts]: I don't see a test for the `TNodeType.IcuContainer` in the series of node type checks. If I recall properly ICU containers "behave" similarly to `<ng-container>`s. I'm not sure if we can bump into ICU containers here, but:
* if no, we should add an assert;
* if yes, we should add a test + handle this case.

File: packages/core/src/render3/view_ref.ts [hunk 3]
```diff
@@ -10,11 +10,13 @@ import {ApplicationRef} from '../application_ref';
...
 import {EmbeddedViewRef as viewEngine_EmbeddedViewRef, InternalViewRef as viewEngine_InternalViewRef} from '../linker/view_ref';
+import {assertDefined} from '../util/assert';
 
 import {checkNoChangesInRootView, checkNoChangesInternal, detectChangesInRootView, detectChangesInternal, markViewDirty, storeCleanupFn} from './instructions/shared';
+import {CONTAINER_HEADER_OFFSET} from './interfaces/container';
 import {TElementNode, TNode, TNodeType, TViewNode} from './interfaces/node';
-import {CONTEXT, FLAGS, HOST, LView, LViewFlags, T_HOST} from './interfaces/view';
-import {destroyLView, renderDetachView} from './node_manipulation';
+import {CONTEXT, FLAGS, HOST, LView, LViewFlags, TView, T_HOST} from './interfaces/view';
+import {destroyLView, getLContainer, renderDetachView} from './node_manipulation';
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/view_ref.ts]: I don't see the new `getLContainer ` import used anywhere in the changed code. Remove?

File: packages/core/src/render3/view_ref.ts (hunk 2 above)
Comment (pkozlowski-opensource) [reply packages/core/src/render3/view_ref.ts]: Also, I would add an assert to verify that we don't bump into `tNode` being `TNodeType.View` (just for completeness, to make sure that we are not missing handling of any TNode type).

File: packages/core/test/acceptance/view_container_ref_spec.ts [hunk 4]
```diff
@@ -872,6 +872,60 @@ describe('ViewContainerRef', () => {
...
+              this.viewContainerRef.createEmbeddedView(this.template, new MyDirEmbedded());
+          this.viewRef.detectChanges();
+        }
+      }
+
+      @Component({
+        selector: 'edit-form',
+        template: `
+          <ng-template [dir]="true">
+            <div *ngIf="true">Text</div>
```
Comment (pkozlowski-opensource) [inline packages/core/test/acceptance/view_container_ref_spec.ts]: Could we add few more test cases:
- nested `<ng-container>`s at the root of an embedded view;
- ICU inside `<ng-container>` - I'm not sure if we can have them at the root of a view;
- a directive that asks for a `ViewContainerRef` at the root of a view (you've got a case where a VCRef is asked on a container);

**Review by pkozlowski-opensource — CHANGES_REQUESTED** [review]
Did the first pass. The direction LGTM but I'm not sure if we are handling all the cases of nodes that can be at the root of a view (left comments enumerating those cases). As the very minimum we need more tests.

File: packages/core/src/render3/view_ref.ts (hunk 1 above)
Comment (mhevery) [reply packages/core/src/render3/view_ref.ts]: fixed

File: packages/core/src/render3/view_ref.ts (hunk 3 above)
Comment (mhevery) [reply packages/core/src/render3/view_ref.ts]: removed

File: packages/core/src/render3/view_ref.ts (hunk 2 above)
Comment (mhevery) [reply packages/core/src/render3/view_ref.ts]: - I have added asserts
- Added tests and was not able to get `Icu` node to show up here.

File: packages/core/test/acceptance/view_container_ref_spec.ts (hunk 4 above)
Comment (mhevery) [reply packages/core/test/acceptance/view_container_ref_spec.ts]: added test cases

**Review by mhevery — COMMENTED** [review]
Updated, PTAL

File: packages/core/src/render3/view_ref.ts [hunk 5]
```diff
@@ -299,27 +302,39 @@ export class RootViewRef<T> extends ViewRef<T> {
...
-    if (tNodeChild.type === TNodeType.ElementContainer) {
-      collectNativeNodes(lView, tNodeChild, result);
-    } else if (tNodeChild.type === TNodeType.Projection) {
+
+    if (tNode.type === TNodeType.ElementContainer) {
+      collectNativeNodes(lView, tNode.child, result);
+    } else if (tNode.type === TNodeType.Container) {
+      const lContainer = lView[tNode.index];
+      const containerTView = tNode.tViews !as TView;
+      ngDevMode && assertDefined(containerTView, 'TView expected');
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/view_ref.ts]: I don't think that this assert is correct for 2 reasons:
* it refers to a `TView` _defined_ by this `<ng-template>` and not _inserted_ into a `ViewContainerRef` asked on this `<ng-template>`;
* many different types of TViews can be inserted into one container so we can't assume that all _inserted_ views share the same `TView`
* it can be `null` when the definition is empty `<ng-template></ng-template>`, AFAIK

There is a confusion between an embedded view _defintion_ and the the _inserted instance_ here.

File: packages/core/test/acceptance/view_container_ref_spec.ts [hunk 6]
```diff
@@ -872,6 +872,65 @@ describe('ViewContainerRef', () => {
...
+            <ng-container>container-content</ng-container>
+            <div i18n>assert TNodeType.IcuContainer is never a root node</div>
```
Comment (pkozlowski-opensource) [inline packages/core/test/acceptance/view_container_ref_spec.ts]: This is not enough, you need to have an actual ICU expression inside an element with the `i18n` attribute. Also, having `div` at the root is not enough as a `<div>` is a root node so no need to descend any further.  An example way of testing it would be `<ng-container i18n>Updated {minutes, plural, =0 {just now} =1 {one minute ago} other {several minutes ago}}</ng-container>` and here we need to descent into the ICU container.

tl;dr; I still believe that we need to handle the case of a ICU container.

File: packages/core/src/render3/view_ref.ts [hunk 7]
```diff
@@ -299,27 +302,39 @@ export class RootViewRef<T> extends ViewRef<T> {
...
-function collectNativeNodes(lView: LView, parentTNode: TNode, result: any[]): any[] {
-  let tNodeChild = parentTNode.child;
-
-  while (tNodeChild) {
-    const nativeNode = getNativeByTNodeOrNull(tNodeChild, lView);
+function collectNativeNodes(lView: LView, tNode: TNode | null, result: any[]): any[] {
+  while (tNode) {
+    ngDevMode && assertNodeOfPossibleTypes(
+                     tNode, TNodeType.Element, TNodeType.Container, TNodeType.Projection,
+                     TNodeType.ElementContainer);
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/view_ref.ts]: I believe that we need to handle a ICU container, see: https://github.com/angular/angular/pull/33457/files#r340598750

File: packages/core/test/acceptance/view_container_ref_spec.ts (hunk 4 above)
Comment (pkozlowski-opensource) [reply packages/core/test/acceptance/view_container_ref_spec.ts]: I haven't seen a test for the case where a `ViewContainerRef ` is injected on a regular element.

**Review by pkozlowski-opensource — CHANGES_REQUESTED** [review]
@mhevery I still believe that there are tests / use-cases missing. How about we review / merge https://github.com/angular/angular/pull/33493 instead that has all the test I could think of (+ fixes for those)?

**Comment by mhevery** [comment]
Closing in as @pkozlowski-opensource took over https://github.com/angular/angular/pull/33493
